File: clients/python/uiform/resources/secrets/external_api_keys.py
import os
from typing import List

from ..._resource import AsyncAPIResource, SyncAPIResource
from ...types.ai_models import AIProvider
from ...types.secrets.external_api_keys import ExternalAPIKey, ExternalAPIKeyRequest
from ...types.standards import PreparedRequest


class ExternalAPIKeysMixin:
    def prepare_create(self, provider: AIProvider, api_key: str) -> PreparedRequest:
        data = {"provider": provider, "api_key": api_key}
        request = ExternalAPIKeyRequest.model_validate(data)
        return PreparedRequest(method="POST", url="/v1/secrets/external_api_keys", data=request.model_dump(mode="json"))

    def prepare_get(self, provider: AIProvider) -> PreparedRequest:
        return PreparedRequest(method="GET", url=f"/v1/secrets/external_api_keys/{provider}")

    def prepare_list(self) -> PreparedRequest:
        return PreparedRequest(method="GET", url="/v1/secrets/external_api_keys")

    def prepare_delete(self, provider: AIProvider) -> PreparedRequest:
        return PreparedRequest(method="DELETE", url=f"/v1/secrets/external_api_keys/{provider}")


class ExternalAPIKeys(SyncAPIResource, ExternalAPIKeysMixin):
    """External API Keys management wrapper"""

    def create(self, provider: AIProvider, api_key: str) -> dict:
        """Add or update an external API key.

        Args:
            provider: The API provider (openai, gemini, anthropic, xai)
            api_key: The API key to store

        Returns:
            dict: Response indicating success
        """

        request = self.prepare_create(provider, api_key)
        response = self._client._prepared_request(request)
        return response

    def get(
        self,
        provider: AIProvider,
    ) -> ExternalAPIKey:
        """Get an external API key configuration.

        Args:
            provider: The API provider to get the key for

        Returns:
            ExternalAPIKey: The API key configuration
        """
        request = self.prepare_get(provider)
        response = self._client._prepared_request(request)
        return response

        return ExternalAPIKey.model_validate(response)

    def list(self) -> List[ExternalAPIKey]:
        """List all configured external API keys.

        Returns:
            List[ExternalAPIKey]: List of API key configurations
        """
        request = self.prepare_list()
        response = self._client._prepared_request(request)

        return [ExternalAPIKey.model_validate(key) for key in response]

    def delete(self, provider: AIProvider) -> dict:
        """Delete an external API key configuration.

        Args:
            provider: The API provider to delete the key for

        Returns:
            dict: Response indicating success
        """
        request = self.prepare_delete(provider)
        response = self._client._prepared_request(request)

        return response


class AsyncExternalAPIKeys(AsyncAPIResource, ExternalAPIKeysMixin):
    """External API Keys management wrapper"""

    async def create(self, provider: AIProvider, api_key: str) -> dict:
        request = self.prepare_create(provider, api_key)
        response = await self._client._prepared_request(request)
        return response

    async def get(self, provider: AIProvider) -> ExternalAPIKey:
        request = self.prepare_get(provider)
        response = await self._client._prepared_request(request)
        return response

    async def list(self) -> List[ExternalAPIKey]:
        request = self.prepare_list()
        response = await self._client._prepared_request(request)
        return [ExternalAPIKey.model_validate(key) for key in response]

    async def delete(self, provider: AIProvider) -> dict:
        request = self.prepare_delete(provider)
        response = await self._client._prepared_request(request)
        return response

File: clients/node/src/resources/secrets/external_api_keys.ts
import { SyncAPIResource, AsyncAPIResource } from '../../resource.js';
import { PreparedRequest } from '../../types/standards.js';
import { AIProvider } from '../../types/ai_models.js';
import { ExternalAPIKey, ExternalAPIKeyRequest } from '../../types/secrets/external_api_keys.js';

export class ExternalAPIKeysMixin {
  prepareCreate(provider: AIProvider, api_key: string): PreparedRequest {
    const request: ExternalAPIKeyRequest = {
      provider,
      api_key,
    };
    
    return {
      method: 'POST',
      url: '/v1/secrets/external_api_keys',
      data: request,
    };
  }

  prepareGet(provider: AIProvider): PreparedRequest {
    return {
      method: 'GET',
      url: `/v1/secrets/external_api_keys/${provider}`,
    };
  }

  prepareList(): PreparedRequest {
    return {
      method: 'GET',
      url: '/v1/secrets/external_api_keys',
    };
  }

  prepareDelete(provider: AIProvider): PreparedRequest {
    return {
      method: 'DELETE',
      url: `/v1/secrets/external_api_keys/${provider}`,
    };
  }
}

export class ExternalAPIKeys extends SyncAPIResource {
  private mixin = new ExternalAPIKeysMixin();

  /**
   * Add or update an external API key.
   *
   * @param provider - The API provider (OpenAI, Anthropic, Gemini, xAI)
   * @param api_key - The API key to store
   * @returns Response indicating success
   */
  create(provider: AIProvider, api_key: string): Promise<Record<string, any>> {
    const request = this.mixin.prepareCreate(provider, api_key);
    const response = this._client._preparedRequest(request);
    return response as Promise<Record<string, any>>;
  }

  /**
   * Get an external API key configuration.
   *
   * @param provider - The API provider to get the key for
   * @returns The API key configuration
   */
  get(provider: AIProvider): Promise<ExternalAPIKey> {
    const request = this.mixin.prepareGet(provider);
    const response = this._client._preparedRequest(request);
    return response as Promise<ExternalAPIKey>;
  }

  /**
   * List all configured external API keys.
   *
   * @returns List of API key configurations
   */
  list(): Promise<ExternalAPIKey[]> {
    const request = this.mixin.prepareList();
    const response = this._client._preparedRequest(request);
    return response as Promise<ExternalAPIKey[]>;
  }

  /**
   * Delete an external API key configuration.
   *
   * @param provider - The API provider to delete the key for
   * @returns Response indicating success
   */
  delete(provider: AIProvider): Promise<Record<string, any>> {
    const request = this.mixin.prepareDelete(provider);
    const response = this._client._preparedRequest(request);
    return response as Promise<Record<string, any>>;
  }
}

export class AsyncExternalAPIKeys extends AsyncAPIResource {
  private mixin = new ExternalAPIKeysMixin();

  /**
   * Add or update an external API key asynchronously.
   */
  async create(provider: AIProvider, api_key: string): Promise<Record<string, any>> {
    const request = this.mixin.prepareCreate(provider, api_key);
    const response = await this._client._preparedRequest(request);
    return response as Record<string, any>;
  }

  /**
   * Get an external API key configuration asynchronously.
   */
  async get(provider: AIProvider): Promise<ExternalAPIKey> {
    const request = this.mixin.prepareGet(provider);
    const response = await this._client._preparedRequest(request);
    return response as ExternalAPIKey;
  }

  /**
   * List all configured external API keys asynchronously.
   */
  async list(): Promise<ExternalAPIKey[]> {
    const request = this.mixin.prepareList();
    const response = await this._client._preparedRequest(request);
    return response as ExternalAPIKey[];
  }

  /**
   * Delete an external API key configuration asynchronously.
   */
  async delete(provider: AIProvider): Promise<Record<string, any>> {
    const request = this.mixin.prepareDelete(provider);
    const response = await this._client._preparedRequest(request);
    return response as Record<string, any>;
  }
}